src/protocols/secure_channel/CASESession.cpp

```diff
@@ -1120,33 +1289,104 @@ CHIP_ERROR CASESession::SetEffectiveTime(void)
     return ASN1ToChipEpochTime(effectiveTime, mValidContext.mEffectiveTime);
 }
 
-CHIP_ERROR CASESession::HandleErrorMsg(const System::PacketBufferHandle & msg)
+void CASESession::OnSuccessStatusReport()
 {
-    ReturnErrorCodeIf(msg->Start() == nullptr || msg->DataLength() != sizeof(SigmaErrorMsg), CHIP_ERROR_MESSAGE_INCOMPLETE);
+    ChipLogProgress(SecureChannel, "Success status report received. Session was established");
+    mPairingComplete = true;
 
-    static_assert(sizeof(SigmaErrorMsg) == sizeof(uint8_t),
-                  "Assuming size of SigmaErrorMsg message is 1 octet, so that endian-ness conversion is not needed");
+    // Forget our exchange, as no additional messages are expected from the peer
+    mExchangeCtxt = nullptr;
 
-    SigmaErrorMsg * pMsg = reinterpret_cast<SigmaErrorMsg *>(msg->Start());
-    ChipLogError(SecureChannel, "Received error (%d) during CASE pairing process", pMsg->error);
+    // Call delegate to indicate pairing completion
+    mDelegate->OnSessionEstablished();
 
+    mState = kInitialized;
+}
+
+CHIP_ERROR CASESession::OnFailureStatusReport(Protocols::SecureChannel::GeneralStatusCode generalCode, uint16_t protocolCode)
+{
     CHIP_ERROR err = CHIP_NO_ERROR;
-    switch (pMsg->error)
+    switch (protocolCode)
     {
-    case SigmaErrorType::kUnsupportedVersion:
-        err = CHIP_ERROR_UNSUPPORTED_CASE_CONFIGURATION;
+    case kProtocolCodeInvalidParam:
+        err = CHIP_ERROR_INVALID_CASE_PARAMETER;
         break;
 
-    case SigmaErrorType::kInvalidSignature:
-    case SigmaErrorType::kInvalidResumptionTag:
-    case SigmaErrorType::kUnexpected:
-        err = CHIP_ERROR_INVALID_CASE_PARAMETER;
+    case kProtocolCodeNoSharedRoot:
+        err = CHIP_ERROR_CERT_NOT_TRUSTED;
         break;
 
     default:
         err = CHIP_ERROR_INTERNAL;
         break;
     };
+    mState = kInitialized;
+    ChipLogError(SecureChannel, "Received error (protocol code %d) during pairing process. %s", protocolCode, ErrorStr(err));
+    return err;
+}
+
+CHIP_ERROR CASESession::ParseSessionResumptionRequest(const System::PacketBufferHandle & msg, bool & resumptionRequested,
+                                                      MutableByteSpan & resumptionID, MutableByteSpan & resume1MIC)
+{
+    CHIP_ERROR err = CHIP_NO_ERROR;
+
+    System::PacketBufferTLVReader tlvReader;
+    TLV::TLVType containerType = TLV::kTLVType_Structure;
+
+    constexpr uint8_t kResumptionIDTag = 6;
+    constexpr uint8_t kResume1MICTag   = 7;
+
+    bool resumptionIDTagFound = false;
+    bool resume1MICTagFound   = false;
+
+    System::PacketBufferHandle msg_clone = msg.CloneData();
+
+    tlvReader.Init(std::move(msg_clone));
+    SuccessOrExit(err = tlvReader.Next(containerType, TLV::AnonymousTag));
+    SuccessOrExit(err = tlvReader.EnterContainer(containerType));
+
+    while (!resumptionIDTagFound || !resume1MICTagFound)
```

Comment:
It's only legal for the resumption tag to be in exact tag order such that there are no lesser tags before. There should be a check through the loop that:

- Each tag is higher than the previous one
- There is no duplicate tags
- MIC is after Resumption  ID (may be covered by previous condition if done as a first step altogether).